nemo/collections/audio/modules/ssl_pretrain_masking.py

```diff
@@ -0,0 +1,88 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import random
+
+import einops
+import torch
+
+from nemo.core.classes import NeuralModule, typecheck
+from nemo.core.neural_types import LengthsType, NeuralType, SpectrogramType
+
+__all__ = ['SSLPretrainWithMaskedPatch']
+
+
+class SSLPretrainWithMaskedPatch(NeuralModule):
+    """
+    Zeroes out fixed size time patches of the spectrogram.
+    All samples in batch are guaranteed to have the same amount of masked time steps.
+
+    Args:
+        patch_size (int): up to how many time steps does one patch consist of.
+            Defaults to 48.
+        mask_fraction (float): how much fraction in each sample to be masked (number of patches is rounded up).
+            Range from 0.0 to 1.0. Defaults to 0.7.
+    """
+
+    @property
+    def input_types(self):
+        """Returns definitions of module input types"""
+        return {
+            "input_spec": NeuralType(('B', 'C', 'D', 'T'), SpectrogramType()),
+            "length": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    @property
+    def output_types(self):
+        """Returns definitions of module output types"""
+        return {"augmented_spec": NeuralType(('B', 'C', 'D', 'T'), SpectrogramType())}
+
+    def __init__(
+        self,
+        patch_size: int = 10,
+        mask_fraction: float = 0.7,
+    ):
+        super().__init__()
+        self.patch_size = patch_size
+        if mask_fraction > 1.0 or mask_fraction < 0.0:
+            raise ValueError('mask_patches cannot be negative')
+        else:
+            self.mask_fraction = mask_fraction
+
+    @typecheck()
+    def forward(self, input_spec, length):
+        augmented_spec = input_spec
+
+        min_len = torch.min(length)
+        if self.training:
+            len_fraction = int(min_len * self.mask_fraction)
+            mask_patches = len_fraction // self.patch_size + int(len_fraction % self.patch_size != 0)
+
+            if min_len < self.patch_size * mask_patches:
+                mask_patches = min_len // self.patch_size
+
+            for idx in range(input_spec.shape[0]):
+                cur_len = length[idx]
+                patches = range(cur_len // self.patch_size)
+                masked_patches = random.sample(patches, mask_patches)
+                for mp in masked_patches:
+                    augmented_spec[idx, :, :, mp * self.patch_size : (mp + 1) * self.patch_size] = 0.0
+        else:
```

Comment:
This is for the validation purpose. In the pretraining stage we still would like to have a way to evaluate the checkpoints and avoid overfitting. In the validation stage (`self.training == False`) we fix the masking pattern instead of generating it randomly.